grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -51,14 +60,546 @@ class ApiGetOpenApiDescriptionHandler(api_call_handler_base.ApiCallHandler):
+                    f"which is not a protobuf message Descriptor, "
+                    f"nor an EnumDescriptor, nor a primitive type.")
+
+  def _ExtractSchemas(self):
+    """Extracts OpenAPI schemas for all the used protobuf types."""
+
+    self.schema_objs = dict()  # Holds OpenAPI representations of types.
+    self._AddPrimitiveTypesSchemas()
+
+    visiting = set()  # Holds state of types extraction (white/gray nodes).
+    router_methods = self.router.__class__.GetAnnotatedMethods()
+    for method_metadata in router_methods.values():
+      args_type = method_metadata.args_type
+      if args_type:
+        if inspect.isclass(args_type) \
+            and issubclass(args_type, rdf_structs.RDFProtoStruct):
+          self._ExtractSchema(args_type.protobuf.DESCRIPTOR, visiting)
+        else:
+          self._ExtractSchema(args_type, visiting)
+
+      result_type = method_metadata.result_type
+      if result_type:
+        if inspect.isclass(result_type) \
+            and issubclass(result_type, rdf_structs.RDFProtoStruct):
+          self._ExtractSchema(result_type.protobuf.DESCRIPTOR, visiting)
+        else:
+          self._ExtractSchema(result_type, visiting)
+
+  def _SetComponents(self):
+    if self.schema_objs is None:
+      raise ValueError("Called _SetComponents before extracting schemas.")
+
+    # The Components Object "components" of the root OpenAPI object.
+    components_obj = dict()
+    schemas_obj = dict()
+
+    type_names = set(self.schema_objs.keys())
+    primitive_types_names = set(self.primitive_types_names)
+    # Create components only for composite types.
+    for type_name in type_names - primitive_types_names:
+      schemas_obj[type_name] = self.schema_objs[type_name]
+
+    components_obj["schemas"] = schemas_obj
+
+    self.open_api_obj["components"] = components_obj
+
+  def _GetSchemaOrReferenceObject(self, type_name: str, is_array=False):
+    """Returns a Schema Object if primitive type, else a Reference Object.
+
+    Primitive, not composite types don't have an actual schema, but rather an
+    equivalent OpenAPI representation that gets returned for them.
+    More complex types are expected to have been previously defined as OpenAPI
+    components and are used through OpenAPI references.
+    """
+    if self.schema_objs is None:
+      raise ValueError("Called _GetSchemaOrReferenceObject before extracting "
+                       "schemas.")
+
+    schema_or_ref_obj = None
+    if type_name in self.primitive_types_names:
+      schema_obj = self.schema_objs[type_name]
+      schema_or_ref_obj = schema_obj
+    else:
+      reference_obj = dict()
+      reference_obj["$ref"] = f"#/components/schemas/{type_name}"
+      schema_or_ref_obj = reference_obj
+
+    if is_array:
+      schema_or_ref_obj = {
+        "type": "array",
+        "items": schema_or_ref_obj
+      }
+
+    return schema_or_ref_obj
+
+  def _GetParameters(self, path_params, query_params):
```

Comment:
Missing docstring (can be one line, but method is quite long so our internal linter will surely complain). Here and in all methods below.

Reply:
Thank you! I added the missing docstring parts and new docstrings to methods that didn't have them.